rfcs/text/0013_saved_object_migrations.md

```diff
@@ -0,0 +1,511 @@
+new index will be in an outdated format. This could cause queries on the data
+to only return a subset of the results which leads to incorrect results or
+silent data loss.
+
+Any in-progress writes such as a bulk operation will continue to run even
+after an index was set to read-only or an alias had been removed. There are
+also certain failure scenarios under which ClusterState (such as aliases or
+index settings) remains inconsistent between nodes for an unbounded
+amount of time. Since Kibana cannot rely on Elasticsearch scripting support
+being enabled, we don't have a mechanism to guarantee that there would be no
+data loss. Instead, we aim to minimize it as much as possible.
+
+To minimize data loss we need to implement a “guard” which will prevent old
+nodes and new nodes from writing at the same time. To perform a rollback for a
+failed migration, this guard will have to be removed. By locating this state
+in the backup snapshot taken before an upgrade, rolling back to the previous
+snapshot will automatically remove the guard.
+
+**7.x Minimize data loss when creating new indices**
+
+Preventing data loss from mixed 7.x versions, introduces two additional steps
+to rollback to a previous version without a snapshot:
+1. (existing) Point the `.kibana` alias to the previous Kibana index `.kibana_n-1`
+2. (existing) Delete `.kibana_n`
+3. (new) Enable writes on `.kibana_n-1`
+4. (new) Delete the dummy "version lock" document from `.kibana_n-1`
+
+Since our documentation and server logs have implicitly encouraged users to
+rollback without using snapshots, many users might have to rely on these
+additional migration steps to perform a rollback. Since even the existing
+steps are error prone, introducing more steps will likely introduce more
+problems than what it solves.
+
+<details>
+  <summary><s>Minimize data loss with mixed 7.x versions</s></summary>
+  
+  1. All future versions of Kibana 7.x will use the `.kibana_saved_objects`
+   alias to locate the current index. If `.kibana_saved_objects` doesn't
+   exist, newer versions will fallback to reading `.kibana`.
+  2. All future versions of Kibana will locate the index that
+     `.kibana_saved_objects` points to and then read and write directly from
+     the _index_ instead of the alias.
+  3. Before starting a migration:
+    1. Write a new dummy "version lock" document to the `.kibana` index with a
+      `migrationVersion` set to the current version of Kibana. If an outdated
+      node is started up after a migration was started it will detect that
+      newer documents are present in the index and refuse to start up.
+    2. Set the outdated index to read-only. Since `.kibana` is never advanced,
+      it will be pointing to a read-only index which prevent writes from
+      6.8+ releases which are already online.
+</details>
+
+## 4.5 Changes in 8.0
+1. Migrations will be in-place and re-use the same index for minor upgrades.
```

Comment:
for upgrades in minor and patch releases?

Reply:
Yes, will clarify.